### PR TITLE
Add Channel{T}(f::Function) constructors.

### PR DESCRIPTION
**Pull request by NHDaly**
The existing Channel constructors are a bit inconvenient to work with. I love that you can create a Channel and spawn a task at the same time; it's super convenient! But the task-spawning constructor is a bit more cumbersome than the non-task-spawning constructor, because it takes its params as keyword arguments instead of the nicer way that the non-task constructor does it.

This PR adds two convenience constructors to combine those styles, to allow you to write things like this:
```julia
Channel{Char}(1) do chnl
    for c in "hello world"
        put!(chnl, c)
    end
end
``` 

----------

The Channel constructors that do and don't create a Task are currently
written in different styles:

 - The non-Task constructor takes a Type {T} as a type parameter, and
 the size as a positional argument.
    - `Channel{Int}(Inf)`
    - `Channel(0)`

 - The Task constructor takes a Function, but the size and type are
keyword arguments:
    - `Channel(c->put!(c,0), ctype=Int, csize=0)`
    - `Channel(ctype=Int, csize=0, taskref=t) do c; put!(c,0); end`

This commit adds convenience methods to the Task-creating constructor,
allowing you to specify the Type and (optionally) size as in the non-Task constructor:

- `Channel{Char}(c->put!(c, 'a'))`
- `Channel{Char}(1) do c; put!(c, 'a'); end`
- `Channel{Char}(csize=2, taskref=t) do c; put!(c, 'a'); end`
- `Channel{Char}(c->put!(c, 'a'), Inf)`

It also adds default constructors, defaulting to unbuffered channels (ie `sz=0`):
- `Channel() = Channel(0)`
- `Channel{T}() where T = Channel{T}(0)`

---

I also added tests for the existing Task-creating constructors, which weren't explicitly tested before.

--------------

EDIT: Some other things to consider:

- [x] In this PR, I kept it so that the empty constructor, `Channel()` is not valid. I think that's probably still reasonable, although one could argue that this should simply use all the defaults -- ie `Channel() = Channel{Any}(0)`. But I dunno, it's probably fine as-is.
- [x] Let me know if the tests are reasonable.

### CODE REVIEW

File: base/channels.jl [hunk 1]
```diff
@@ -95,6 +95,22 @@ take!(c) = "Hello"
 julia> istaskdone(taskref[])
 true
 ```
+
```
Comment (vchuravy) [inline base/channels.jl]: I think there is a `!!! note` for stating the Julia version since when a change is valid (this will be 1.2)

**Review by vchuravy — APPROVED** [review]
That looks like a good enhancement

File: base/channels.jl (hunk 1 above)
Comment (fredrikekre) [reply base/channels.jl]: `!!! compat "Julia 1.2"`

Comment (NHDaly) [reply base/channels.jl]: so just add this at the bottom of the comment, right here?:
```suggestion
!!! compat "Julia 1.2"
```

**Comment by vchuravy** [comment]
Defaulting to unbuffered channels seems right to me.

File: base/channels.jl (hunk 1 above)
Comment (fredrikekre) [reply base/channels.jl]: Yea, see the note in CONTRIBUTING.md and other examples in the code base.

Comment (NHDaly) [reply base/channels.jl]: Thanks @fredrikekre. Sorry this took me so long to do.

I've added a compat note for these new constructors and the new default constructor that I just added, `Channel()`.

**Comment by NHDaly** [comment]
PTAL: This should be good to review again!

I have added compat statements for julia 1.3, and set the default `Channel()` constructor to default to unbuffered channels of `Any`.

**Comment by NHDaly** [comment]
Please take another look 😄 

I know that feature-freeze for 1.3 is coming up soon, and I feel like since this is going to be the multithreading release, it would be nice to have a user-friendly interface on Channels.

Hopefully this can be resolved before feature freeze? :) Thanks!

**Comment by vchuravy** [comment]
Ah sorry should have been squashed -- mea culpa.

File: base/channels.jl [hunk 2]
```diff
@@ -105,6 +129,12 @@ function Channel(func::Function; ctype=Any, csize=0, taskref=nothing)
     isa(taskref, Ref{Task}) && (taskref[] = task)
     return chnl
 end
+function Channel{T}(f::Function, sz=0) where T
```
Comment (JeffBezanson) [inline base/channels.jl]: This causes a method overwrite warning. This signature can be `Channel{T}(f::Function, sz)`.

Comment (NHDaly) [reply base/channels.jl]: Oops, sorry! Fixed in https://github.com/JuliaLang/julia/pull/32818

Thanks, good catch! Sorry I didn't notice (the error comes up while building and it's easy to miss. I'll keep an eye out for that in the future!) :)

**Pull request merged**